flips/20211208-cadence-remove-optional-reference.md

```diff
@@ -0,0 +1,116 @@
+# Optional References to Indexed Accesses
+
+| Status        | (Approved)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [722](https://github.com/onflow/flow/pull/722) |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2021-12-09                                           |
+
+## Objective
+
+This removes the ability to create a reference to an optional value in Cadence by 
+creating a reference to the result of an indexed access. This was previously the 
+only way to obtain a reference to an optional, and as such this change effectively 
+removes these references from the language entirely. Instead, it allows users to 
+make the type on the right-hand side of an `as` expression an optional, and 
+allows users to create optionally-typed references in place of references
+to optional values. 
+
+## Motivation
+
+As detailed in https://github.com/onflow/cadence/issues/747, it is possible to 
+crash execution with a failed dereference that is not caught by the typechecker
+by creating a reference to the result of an indexed access, and then accessing
+a field on that reference. This occurs because while references to optional 
+values are normally disallowed, references to indexed accesses are special cased
+in the checker. In order to proceed with a stable release of Cadence, this feature
+needs to be removed. 
+
+## User Benefit
+
+This removes an unsound behavior from the type checker, preventing an error
+from occurring at runtime. 
```

Comment:
The current feature is not unsound behaviour, it was intentionally added. At the time it was added there were two options, references to optionals and optional references, and we opted for the current former. The proposal suggests the other option, optional references.

The user benefit is still true: The proposed behaviour removes the potential for the program to unexpectedly abort due to the implicit unwrap of the references optional failing.

Reply:
Is this not still unsound behavior, even if it is intentional? This makes it possible to obtain a value that has a runtime type that differs from its static type, which feels like a pretty straightforward example of unsoundness.